### PR TITLE
boards: arm: add support for NUCLEO-L011K4 board

### PR DESCRIPTION
**Pull request by sdaglish**
Support for the NUCLEO-L011K4 development board with STM32L011K4 SoC.

Although the SoC only contains 16K flash and 2K RAM, it has been tested
on a number of samples and has worked as expected.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

	new file:   boards/arm/nucleo_l011k4/CMakeLists.txt
	new file:   boards/arm/nucleo_l011k4/Kconfig.board
	new file:   boards/arm/nucleo_l011k4/Kconfig.defconfig
	new file:   boards/arm/nucleo_l011k4/board.cmake
	new file:   boards/arm/nucleo_l011k4/doc/img/nucleo_l011k4.jpg
	new file:   boards/arm/nucleo_l011k4/doc/index.rst
	new file:   boards/arm/nucleo_l011k4/nucleo_l011k4.dts
	new file:   boards/arm/nucleo_l011k4/nucleo_l011k4.yaml
	new file:   boards/arm/nucleo_l011k4/nucleo_l011k4_defconfig
	new file:   boards/arm/nucleo_l011k4/pinmux.c
	new file:   boards/arm/nucleo_l011k4/support/openocd.cfg

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition!

Some initial changes should be done on the PR to confirm with zephyr criterias:
- Please squash all commits into a single one (we don't use merge commits, all commits will be put on top of the branch, so we expect one 'clean' commit)
- In commit message, nothing should appear after the sign-off

**Comment by sdaglish** [comment]
> 
> 
> Thanks for this addition!
> 
> Some initial changes should be done on the PR to confirm with zephyr criterias:
> 
>     * Please squash all commits into a single one (we don't use merge commits, all commits will be put on top of the branch, so we expect one 'clean' commit)
> 
>     * In commit message, nothing should appear after the sign-off

Commits have now been squashed and lines after sign-off removed from commit.

File: boards/arm/nucleo_l011k4/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,170 @@
+.. _nucleo_l011k4_board:
+
+ST Nucleo L011K4
+################
+
+Overview
+********
+The STM32 Nucleo-32 development board with STM32L011K4 MCU, supports Arduino and ST morpho connectivity.
```
Comment (erwango) [inline boards/arm/nucleo_l011k4/doc/index.rst:8]: ST morpho is not supported
Supported Arduino connector is Nano V3

File: boards/arm/nucleo_l011k4/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+The STM32 Nucleo board provides an affordable, and flexible way for users to try out new concepts,
+and build prototypes with the STM32 microcontroller, choosing from the various
+combinations of performance, power consumption, and features.
+
+The Arduino* Uno V3 connectivity support and the ST morpho headers allow easy functionality
```
Comment (erwango) [inline boards/arm/nucleo_l011k4/doc/index.rst:14]: To be upated

File: boards/arm/nucleo_l011k4/arduino_r3_connector.dtsi [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+			   <11 0 &gpiob 6 0>,	/* D5 */
+			   <12 0 &gpiob 1 0>,	/* D6 */
+			   <13 0 &gpioc 14 0>,	/* D7 */
+			   <14 0 &gpioc 15 0>,	/* D8 */
+			   <15 0 &gpioa 8 0>,	/* D9 */
+			   <16 0 &gpioa 11 0>,	/* D10 */
+			   <17 0 &gpiob 5 0>,	/* D11 */
+			   <18 0 &gpiob 4 0>,	/* D12 */
+			   <19 0 &gpiob 3 0>;	/* D13 */
+	};
```
Comment (erwango) [inline boards/arm/nucleo_l011k4/arduino_r3_connector.dtsi:35]: You're describing arduino V3 connector, but board provides arduino Nano.
There might be some compatibility but aim of "connectors" is to provide electronic and physical compatibility, which is not the case here.

File: boards/arm/nucleo_l011k4/nucleo_l011k4.dts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	chosen {
+		zephyr, console = &usart2;
+		zephyr, shell - uart = &usart2;
+		zephyr, sram = &sram0;
+		zephyr, flash = &flash0;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+green_led_2:    led_2 {
```
Comment (erwango) [inline boards/arm/nucleo_l011k4/nucleo_l011k4.dts:25]: Please fix indentation.

File: boards/arm/nucleo_l011k4/nucleo_l011k4.dts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#include <st/l0/stm32l011k(3-4)tx-pinctrl.dtsi>
+#include "arduino_r3_connector.dtsi"
+
+/ {
+	model = "STMicroelectronics STM32L011K4-NUCLEO board";
+	compatible = "st,stm32l011k4-nucleo";
+
+	chosen {
+		zephyr, console = &usart2;
+		zephyr, shell - uart = &usart2;
```
Comment (erwango) [inline boards/arm/nucleo_l011k4/nucleo_l011k4.dts:18]:
```suggestion
		zephyr, shell-uart = &usart2;
```
Please fix in the whole file.

**Review by erwango — CHANGES_REQUESTED** [review]
Some comments mainly around connectors.
Otherwise this is the first PR I see that uses deivcetree pinctrl from the start, which makes me quite happy :-)

File: boards/arm/nucleo_l011k4/pinmux.c [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#include <sys/sys_io.h>
+
+#include <pinmux/stm32/pinmux_stm32.h>
+
+/* pin assignments for NUCLEO-L011K4 board */
+static const struct pin_config pinconf[] = {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(i2c1), okay) && CONFIG_I2C
+	{STM32_PIN_PA4, STM32L0_PINMUX_FUNC_PA4_I2C1_SCL},
+	{STM32_PIN_PA10, STM32L0_PINMUX_FUNC_PA10_I2C1_SDA},
+#endif
```
Comment (erwango) [inline boards/arm/nucleo_l011k4/pinmux.c:19]: This could be moved to device tree as well

File: boards/arm/nucleo_l011k4/doc/index.rst (hunk 1 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/doc/index.rst:8]: Reference to ST morpho removed, and Arduino changed to "Arduino Nano V3"

File: boards/arm/nucleo_l011k4/doc/index.rst (hunk 2 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/doc/index.rst:14]: Updated to remove ST morpho and change Arduino Uno to Arduino Nano

File: boards/arm/nucleo_l011k4/doc/index.rst [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+Hardware
+********
+Nucleo L011K4 provides the following hardware components:
+
+- STM32 microcontroller in LQFP32 package
+- Two types of extension resources:
+
+  - Arduino* Uno V3 connectivity
+  - ST morpho extension pin headers for full access to all STM32 I/Os
```
[the remark this answers is not in the document]
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/doc/index.rst:39]: Updated

File: boards/arm/nucleo_l011k4/arduino_r3_connector.dtsi (hunk 3 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/arduino_r3_connector.dtsi:35]: I need to look into this further. The pins above are for the Arduino Nano compatibility as written on the Nucleo board case.

File: boards/arm/nucleo_l011k4/nucleo_l011k4.dts (hunk 4 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/nucleo_l011k4.dts:25]: Indentation fixed.

File: boards/arm/nucleo_l011k4/nucleo_l011k4.dts (hunk 5 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/nucleo_l011k4.dts:18]: Fixed all instances in file.

File: boards/arm/nucleo_l011k4/nucleo_l011k4.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+type: mcu
+arch: arm
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
+supported:
+  - arduino_gpio
+  - arduino_i2c
+  - arduino_spi
```
[the remark this answers is not in the document]
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/nucleo_l011k4.yaml:12]: Lines removed.

File: boards/arm/nucleo_l011k4/arduino_r3_connector.dtsi (hunk 3 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/arduino_r3_connector.dtsi]: I have removed all reference to Arduino pins.

File: boards/arm/nucleo_l011k4/nucleo_l011k4.dts (hunk 5 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/nucleo_l011k4.dts]: I found a few more formatting errors that occurred after using uncrustify, which have now also been fixed.

File: boards/arm/nucleo_l011k4/pinmux.c (hunk 6 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/pinmux.c]: Moved from pinmux.c to device tree.

**Review by erwango — COMMENTED** [review]
@sdaglish I'm fine with current commit.
Though, I'm worried about the PR state vs CI.
Can you rebase on top of master and push once again?

**Comment by sdaglish** [comment]
> @sdaglish I'm fine with current commit.
> Though, I'm worried about the PR state vs CI.
> Can you rebase on top of master and push once again?

Project rebased on top of master and pushed.

**Review by erwango — APPROVED** [review]

File: boards/arm/nucleo_l011k4/support/openocd.cfg [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# This is an ST NUCLEO-L053R8 board with single STM32L011K4 chip.
```
Comment (ABOSTM) [inline boards/arm/nucleo_l011k4/support/openocd.cfg:1]:
```suggestion
# This is an ST NUCLEO-L011K4 board with single STM32L011K4 chip.
```

File: boards/arm/nucleo_l011k4/support/openocd.cfg [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+# http://www.st.com/en/evaluation-tools/nucleo-l053r8.html
```
Comment (ABOSTM) [inline boards/arm/nucleo_l011k4/support/openocd.cfg:2]:
```suggestion
# http://www.st.com/en/evaluation-tools/nucleo-l011k4.html
```

File: boards/arm/nucleo_l011k4/doc/index.rst [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  - GCC-based IDEs
+
+More information about STM32L011K4 can be found in the
+`STM32L0x1 reference manual`_
+
+
+Supported Features
+==================
+
+The Zephyr nucleo_l053r8 board configuration supports the following hardware features:
```
Comment (ABOSTM) [inline boards/arm/nucleo_l011k4/doc/index.rst:74]:
```suggestion
The Zephyr nucleo_l011k4 board configuration supports the following hardware features:
```

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Few typo

File: boards/arm/nucleo_l011k4/support/openocd.cfg (hunk 9 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/support/openocd.cfg]: Typo fixed.

File: boards/arm/nucleo_l011k4/support/openocd.cfg (hunk 10 above)
Comment (sdaglish) [reply boards/arm/nucleo_l011k4/support/openocd.cfg]: Typo fixed.

**Review by ABOSTM — APPROVED** [review]
LGTM

**Pull request merged**